extensions/amp-position-observer/0.1/amp-position-observer.js

```diff
@@ -0,0 +1,378 @@
+ * limitations under the License.
+ */
+
+import {ActionTrust} from '../../../src/action-trust';
+import {getServiceForDoc} from '../../../src/service';
+import {Services} from '../../../src/services';
+import {createCustomEvent} from '../../../src/event-helper';
+import {isExperimentOn} from '../../../src/experiments';
+import {dev, user} from '../../../src/log';
+import {
+  RelativePositions,
+  layoutRectsRelativePos,
+  layoutRectLtwh,
+} from '../../../src/layout-rect';
+import {
+  Layout,
+  getLengthNumeral,
+  getLengthUnits,
+  assertLength,
+  parseLength,
+} from '../../../src/layout';
+import {
+  installPositionObserverServiceForDoc,
+  PositionObserverFidelity,
+  PositionInViewportEntryDef,
+} from '../../../src/service/position-observer-impl';
+
+const TAG = 'amp-position-observer';
+
+export class AmpVisibilityObserver extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?../../../src/service/action-impl.ActionService} */
+    this.action_ = null;
+
+    /** @private {?Element} */
+    this.scene_ = null;
+
+    /** @private {!boolean} */
+    this.isVisible_ = false;
+
+    /** @private {?../../../src/service/position-observer-impl.AmpDocPositionObserver} */
+    this.positionObserver_ = null;
+
+    /** @private {!number} */
+    this.topRatio_ = 0;
+
+    /** @private {!number} */
+    this.bottomRatio_ = 0;
+
+    /** @private {!string} */
+    this.topMarginExpr_ = '0';
+
+    /** @private {!string} */
+    this.bottomMarginExpr_ = '0';
+
+    /** @private {!number} */
+    this.resolvedTopMargin_ = 0;
+
+    /** @private {!number} */
+    this.resolvedBottomMargin_ = 0;
+
+    /** @private {?../../../src/layout-rect.LayoutRectDef} */
+    this.viewportRect_ = null;
+
+    /** @private {?string} */
+    this.targetSelector_ = null;
+
+    /** @private {!number} */
+    this.scrollProgress_ = 0;
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.NODISPLAY;
+  }
+
+  /** @override */
+  buildCallback() {
+    user().assert(isExperimentOn(this.win, TAG), `${TAG} experiment is off.`);
+    // Since this is a functional component and not visual,
+    // layoutCallback is meaningless. We delay the heavy work until
+    // we become visible.
+    const viewer = Services.viewerForDoc(this.getAmpDoc());
+    viewer.whenFirstVisible().then(this.init_.bind(this));
+  }
+
+  /**
+   * @private
+   */
+  init_() {
+    this.parseAttributes_();
+    this.discoverScene_();
+    this.action_ = Services.actionServiceForDoc(this.element);
+    this.maybeInstallPositionObserver_();
+    this.positionObserver_.observe(this.scene_, PositionObserverFidelity.HIGH,
+        this.positionChanged_.bind(this)
+    );
+  }
+
+  /**
+   * Dispatches the `enter` event.
+   * @private
+   */
+  triggerEnter_() {
+    const name = 'enter';
+    const evt = createCustomEvent(this.win, `${TAG}.${name}`, {});
```

Comment:
done. createCustomEvent needs an event. They also carry the `details` part of the event